File: Traveller/js/NameGenerator.js
(function (global) {
    if (typeof global.JSON == "undefined" || !global.JSON) {
        global.JSON = {};
    }
    if(!global.JSON.minify){
        global.JSON.minify = function JSON_minify(json) {

            var tokenizer = /"|(\/\*)|(\*\/)|(\/\/)|\n|\r/g,
                in_string = false,
                in_multiline_comment = false,
                in_singleline_comment = false,
                tmp, tmp2, new_str = [], ns = 0, from = 0, lc, rc,
                prevFrom
                ;
    
            tokenizer.lastIndex = 0;
    
            while (tmp = tokenizer.exec(json)) {
                lc = RegExp.leftContext;
                rc = RegExp.rightContext;
                if (!in_multiline_comment && !in_singleline_comment) {
                    tmp2 = lc.substring(from);
                    if (!in_string) {
                        tmp2 = tmp2.replace(/(\n|\r|\s)+/g, "");
                    }
                    new_str[ns++] = tmp2;
                }
                prevFrom = from;
                from = tokenizer.lastIndex;
    
                // found a " character, and we're not currently in
                // a comment? check for previous `\` escaping immediately
                // leftward adjacent to this match
                if (tmp[0] == "\"" && !in_multiline_comment && !in_singleline_comment) {
                    // perform look-behind escaping match, but
                    // limit left-context matching to only go back
                    // to the position of the last token match
                    //
                    // see: https://github.com/getify/JSON.minify/issues/64
                    tmp2 = lc.substring(prevFrom).match(/\\+$/);
    
                    // start of string with ", or unescaped " character found to end string?
                    if (!in_string || !tmp2 || (tmp2[0].length % 2) == 0) {
                        in_string = !in_string;
                    }
                    from--; // include " character in next catch
                    rc = json.substring(from);
                }
                else if (tmp[0] == "/*" && !in_string && !in_multiline_comment && !in_singleline_comment) {
                    in_multiline_comment = true;
                }
                else if (tmp[0] == "*/" && !in_string && in_multiline_comment && !in_singleline_comment) {
                    in_multiline_comment = false;
                }
                else if (tmp[0] == "//" && !in_string && !in_multiline_comment && !in_singleline_comment) {
                    in_singleline_comment = true;
                }
                else if ((tmp[0] == "\n" || tmp[0] == "\r") && !in_string && !in_multiline_comment && in_singleline_comment) {
                    in_singleline_comment = false;
                }
                else if (!in_multiline_comment && !in_singleline_comment && !(/\n|\r|\s/.test(tmp[0]))) {
                    new_str[ns++] = tmp[0];
                }
            }
            new_str[ns++] = rc;
            return new_str.join("");
        };
    }    
})(
    // attempt to reference the global object
    typeof globalThis != "undefined" ? globalThis :
        typeof global != "undefined" ? global :
            typeof window != "undefined" ? window :
                typeof self != "undefined" ? self :
                    typeof this != "undefined" ? this :
                        {}
);
export function NameGenerator(sourceJson,callback,forbiddenWords,randomizer,fromObject){
    if(typeof fromObject == "undefined"){
        fromObject = false;
    }
    var MathRandom;
    if(randomizer){
        MathRandom = randomizer;
    }else{
        MathRandom = pseudoRandomNumberGenerator((Math.random()*1000000).toString());
    }    
    function pseudoRandomNumberGenerator(word){
        function xmur3(str) {
            for(var i = 0, h = 1779033703 ^ str.length; i < str.length; i++) {
                h = Math.imul(h ^ str.charCodeAt(i), 3432918353);
                h = h << 13 | h >>> 19;
            } return function() {
                h = Math.imul(h ^ (h >>> 16), 2246822507);
                h = Math.imul(h ^ (h >>> 13), 3266489909);
                return (h ^= h >>> 16) >>> 0;
            }
        }
        function xoshiro128ss(a, b, c, d) {
            return function() {
                var t = b << 9, r = a * 5; r = (r << 7 | r >>> 25) * 9;
                c ^= a; d ^= b;
                b ^= c; a ^= d; c ^= t;
                d = d << 11 | d >>> 21;
                return (r >>> 0) / 4294967296;
            }
        }
        var seed = xmur3(word);
        return xoshiro128ss(seed(), seed(), seed(), seed());
    }
    var defaultForbiddenWords = ["bitch","anus","ass"];
    if(typeof forbiddenWords === "undefined" || forbiddenWords === null){
        forbiddenWords = defaultForbiddenWords;
    }
    this.getRandomName = getRandomName;
    this.setForbiddenWords = setForbiddenWords;
    this.restoreDefaultForbiddenWords = restoreDefaultForbiddenWords;
    this.keys = [];
    this.templates = {};
    this.unpackStringTemplate = unpackStringTemplate;
    var UnpackedStringTemplates = { names: {} };
    this.getRandomNameAndPattern = getRandomNameAndPattern;
    if(fromObject){
        templates = getWeightedPatternsFromJSON({}, sourceJson);
        var keys = Object.keys(templates);
        var generator = {templates:templates,keys:keys,getRandomName:getRandomName,getRandomNameAndPattern:getRandomNameAndPattern, setRandom:setRandom, unpackStringTemplate:unpackStringTemplate, setSeed:setSeed};
        callback(generator);
    }else{
        var xhr = new XMLHttpRequest();
        xhr.open("GET", sourceJson);
        xhr.addEventListener("load", function loadStrings(evt) {
            var rawJson = xhr.responseText;
            templates = getWeightedPatternsFromJSON({}, JSON.parse(JSON.minify(rawJson)));
            var keys = Object.keys(templates);
            var generator = {templates:templates,keys:keys,getRandomName:getRandomName,setRandom:setRandom,getRandomNameAndPattern:getRandomNameAndPattern,unpackStringTemplate:unpackStringTemplate, setSeed:setSeed};
            callback(generator);
        });
        xhr.send();
    }
    
    function setSeed(seed){
        MathRandom = pseudoRandomNumberGenerator(seed);
    }
    function setRandom(random){
        MathRandom = random;
    }
    function setForbiddenWords(words){
        forbiddenWords = words;
    }
    function restoreDefaultForbiddenWords(){
        forbiddenWords = defaultForbiddenWords;
    }
    function getRandomNameAndPattern(key,forbiddenWords){
        var getNameWithChain = getRandomNameAndPattern;
        var boundGetNameWithChain = getNameWithChain.bind(this);
        var result = boundGetNameWithChain(boundGetNameWithChain, key,forbiddenWords);

        return {name:result.piece, pattern:result.templatePiece};

        
        function getRandomNameAndPattern(self, key, bannedWords,depth) {
            if(typeof bannedWords === "undefined"){
                bannedWords = self.forbiddenWords;
                if(typeof forbiddenWords === "undefined" || forbiddenWords === null){
                    bannedWords = [];
                }
            }
            if(typeof depth === "undefined"){
                depth = 1;
            }
            if(key.indexOf("{") == -1){
                var templatesSubset = this.templates[key];
                var templateRoll =(templatesSubset.length * MathRandom()) >>> 0 ;
                var template = templatesSubset[templateRoll];
            }else{
                template = key;
            }     

            var phrases = this.unpackStringTemplate(template); // convert string into array of phrase objects with child segments
            var phraseRoll = (phrases.length * MathRandom()) >>> 0
            var phrase = phrases[ phraseRoll ];
            var currentText = "";
            var currentTemplate = "";
            var references = [];
            var checkNot = -1;
            var capitalize = false;
            
            var phraseContainsAnyReferences = false;
            for (var i = 0, len = phrase.length; i < len; i++) {
                var segment = phrase[i];
                if (segment.reference || segment.references) {
                    phraseContainsAnyReferences = true;
                    break;   
                }
            }
            for (var i = 0, len = phrase.length; i < len; i++) { // process each segment of the selected phrase
                var segment = phrase[i];
                var piece = "";
                var pieceWithChain = {};
                if (segment.text) {
                    piece = segment.text;                
                    if(piece.endsWith(">") && references.length > 0){
                        var matches = piece.match(/<!\d>/g);
                        if(matches != null && matches.length > 0){
                            var number = +(matches[matches.length-1].match(/\d/)[0]);
                            checkNot = number;
                            var regex = new RegExp("<!"+number.toString() + ">","g");
                            piece = piece.replace(regex,"");
                            capitalize = false;
                        }else{
                            checkNot = -1;
                            var matches = piece.match(/<c>/g);
                            if(matches != null && matches.length > 0){
                                capitalize = true;
                                piece = piece.replace(/<c>/g,"");
                            }
                        }
                    }else{
                        checkNot = -1;
                        capitalize = false;
                    }
                    currentText += piece;
                    if(phraseContainsAnyReferences){ currentTemplate += piece; }else{ currentTemplate += "{"+key+"}"; }
                    
                } else if (segment.reference) {
                    pieceWithChain = self(self,segment.reference,bannedWords,depth+1);
                    piece = pieceWithChain.piece;
                    if(checkNot >= 0){
                        while(piece == references[checkNot]){
                            pieceWithChain = self(self,segment.reference,bannedWords,depth+1);
                            piece = pieceWithChain.piece;
                        }
                    }
                    if(capitalize){
                        piece = addCaps(piece);
                    }
                    currentText += piece;
                    currentTemplate += pieceWithChain.templatePiece;
                    references.push(piece);

                } else if (segment.references) {
                    var refIndex = (segment.references.length * MathRandom()) >>> 0;
                    var ref = segment.references[refIndex];
                    pieceWithChain = self(self,ref,bannedWords,depth);
                    piece = pieceWithChain.piece;
                    if(checkNot >= 0){
                        while(piece == references[checkNot]){
                            pieceWithChain = self(self,ref,bannedWords,depth)
                            piece = pieceWithChain.piece;
                        }
                    }
                    if(capitalize){
                        piece = addCaps(piece);
                    }
                    currentText += piece;
                    currentTemplate += pieceWithChain.templatePiece;
                    references.push(piece);
                }
            }
            if(currentText.indexOf("<") > 0 && currentText.indexOf("&") > 0 && currentText.indexOf(">") > 0){
                for(var i = 0, len = references.length; i < len; i++){
                    var regex = new RegExp("\<\&"+i+">","g");
                    currentText = currentText.replace(regex,references[i]);
                }
            }
            if(bannedWords.indexOf(currentText) >= 0){
                var currentTextWithChain = self(self,key, bannedWords,true);
                currentText = currentTextWithChain.piece;
            }
            currentText = currentText.replace(/\s+/g," ");

            return {piece:currentText, templatePiece:currentTemplate};
        }

    }

    


    function getRandomName(key, bannedWords, topLevel) {
        var extraLogs = false;
        if(typeof bannedWords === "undefined"){
            bannedWords = forbiddenWords;
        }
        if(typeof topLevel === "undefined"){
            topLevel = true;
            extraLogs = true;
            //getRandomNameWithKeyChain(key, bannedWords, topLevel, []);
        }
        var templatesSubset = this.templates[key];
        var templateRoll =(templatesSubset.length * MathRandom()) >>> 0 ;
        var template = templatesSubset[templateRoll];
        var phrases = this.unpackStringTemplate(template); // convert string into array of phrase objects with child segments
        // e.g. [{reference:"ship.navalname"}] or [{references:["ship.navalname","system.name"]}] or [{text:"a"}]
        UnpackedStringTemplates.names[key] = phrases;
        var phraseRoll = (phrases.length * MathRandom()) >>> 0
        var phrase = phrases[ phraseRoll ];
        var currentText = "";
        var references = [];
        var checkNot = -1;
        var capitalize = false;
        for (var i = 0, len = phrase.length; i < len; i++) { // process each segment of the selected phrase
            var segment = phrase[i];
            var piece = "";
            if (segment.text) {
                piece = segment.text;                
                if(piece.endsWith(">") && references.length > 0){
                    var matches = piece.match(/<!\d>/g);
                    if(matches != null && matches.length > 0){
                        var number = +(matches[matches.length-1].match(/\d/)[0]);
                        checkNot = number;
                        var regex = new RegExp("<!"+number.toString() + ">","g");
                        piece = piece.replace(regex,"");
                        capitalize = false;
                    }else{
                        checkNot = -1;
                        var matches = piece.match(/<c>/g);
                        if(matches != null && matches.length > 0){
                            capitalize = true;
                            piece = piece.replace(/<c>/g,"");
                        }
                    }
                }else{
                    checkNot = -1;
                    capitalize = false;
                }
                currentText += piece;
            } else if (segment.reference) {
                piece = this.getRandomName(segment.reference,bannedWords,false);
                if(checkNot >= 0){
                    while(piece == references[checkNot]){
                        piece = this.getRandomName(segment.reference,bannedWords,false);
                    }
                }
                if(capitalize){
                    piece = addCaps(piece);
                }
                currentText += piece;
                references.push(piece);
            } else if (segment.references) {
                var ref = segment.references[(segment.references.length * MathRandom()) >>> 0];
                piece = this.getRandomName(ref,bannedWords,false);;
                if(checkNot >= 0){
                    while(piece == references[checkNot]){
                        piece = this.getRandomName(ref,bannedWords,false);
                    }
                }
                if(capitalize){
                    piece = addCaps(piece);
                }
                currentText += piece;
                references.push(piece);
            }
        }
        if(currentText.indexOf("<") > 0 && currentText.indexOf("&") > 0 && currentText.indexOf(">") > 0){
            for(var i = 0, len = references.length; i < len; i++){
                var regex = new RegExp("\<\&"+i+">","g");
                currentText = currentText.replace(regex,references[i]);
            }
        }
        if(bannedWords.indexOf(currentText) >= 0){
            currentText = this.getRandomName(key, bannedWords,true);
        }
        return currentText.replace(/\s+/g," ");
    }
    function addCaps(text) {
        var capitalizedString = "";
        var lastCharacter = null, currCharacter = " ";
        for (var i = 0, len = text.length; i < len; i++) {
            lastCharacter = currCharacter;
            currCharacter = text[i];
            if (lastCharacter === " " || lastCharacter === "-") {
                capitalizedString += currCharacter.toUpperCase();
            } else {
                capitalizedString += currCharacter;
            }
        }
        return capitalizedString;
    }
    function unpackStringTemplate(template) {
        var bracketDepth = 0;
        var phraseArray = [];
        var currentString = "";
        if(template === ""){ return [""];}
        // build outer array
        for (var i = 0, len = template.length; i < len; i++) {
            var character = template[i];
            if (character === "{") {
                bracketDepth += 1;
                currentString += character;
            } else if (character === "}") {
                bracketDepth -= 1;
                currentString += character;
            } else if (character === "|" && bracketDepth === 0) {
                phraseArray.push(currentString);
                currentString = "";
            } else {
                currentString += character;
            }
        }
        if (currentString.length > 0) {
            phraseArray.push(currentString);
        }
        // for each phrase in the array, get the segments
        var isTextSegment = true, isRefSegment = false, isRefArray = false;
        currentString = "", bracketDepth = 0;
        var returnArray = []; // this array holds the phrases
        for (var i = 0, len = phraseArray.length; i < len; i++) {
            var phrase = phraseArray[i];
            var currArr = []; // this array holds the segments for a phrase
            var refArr = []; // this object holds a references segment under construction;
            for (var j = 0, jlen = phrase.length; j < jlen; j++) {
                var character = phrase[j];
                if (character === "{") {
                    if (currentString.length > 0) {
                        if (isTextSegment) {
                            currArr.push({ text: currentString });
                        }
                    }
                    isTextSegment = false;
                    isRefSegment = true;
                    currentString = "";
                } else if (character === "}") {
                    if (isRefSegment) {
                        currArr.push({ reference: currentString });
                        currentString = "";
                        isRefSegment = false;
                        currentString = "";
                    } else if (isRefArray) {
                        refArr.push(currentString);
                        currentString = "";
                        currArr.push({ references: refArr });
                        isRefArray = false;
                        refArr = [];
                    }
                    currentString = "";
                } else if (character === "|") {
                    if (isRefSegment) {
                        isRefArray = true;
                        isRefSegment = false;
                        refArr.push(currentString);
                        currentString = "";
                    } else if (isRefArray) {
                        refArr.push(currentString);
                        currentString = "";
                    }
                } else {
                    if (!isRefSegment && !isRefArray) {
                        isTextSegment = true;
                    }
                    currentString += character;
                }
            }
            if (currentString.length > 0) {
                if (isTextSegment) {
                    currArr.push({ text: currentString });
                    currentString = "";
                }
            }
            returnArray.push(currArr);
        }
        return returnArray;
    }
    function getWeightedPatternsFromJSON(returnObj, obj) {
        for (var key in obj) {
            if (key === "patterns") { continue; }
            var collection = obj[key];
            var collectionIsArray = collection.constructor === Array;
            returnObj[key] = getWeightedPatternsFromJSONSegment(collection);
            if (!collectionIsArray) {
                for (var subkey in collection) {
                    if (subkey === "patterns") { continue; }
                    var subcollection = collection[subkey];
                    returnObj[key + "." + subkey] = getWeightedPatternsFromJSONSegment(subcollection);
                }
            }
        }
        return returnObj;
    }
    function getWeightedPatternsFromJSONSegment(collection) {
        var weightedPatterns = [];
        if (typeof collection.patterns !== "undefined") {

            var patterns = collection.patterns;
            for (var i = 0, len = patterns.length; i < len; i++) {
                var pattern = patterns[i];
                var weight = 1, isString = pattern.constructor === String;
                var patternText = "";
                if (!isString) {
                    if (pattern.weight) {
                        weight = pattern.weight;
                    }
                    if (pattern.format) {
                        patternText = pattern.format
                    }
                } else { // look for shorthand of weight:template in string
                    var indexOfColon = pattern.indexOf(":");
                    if (indexOfColon > 0) {
                        var preColonText = pattern.substring(0, indexOfColon);
                        var number = +(preColonText);
                        if (number >= 0) {
                            weight = number;
                            patternText = pattern.substring(indexOfColon + 1);
                        } else {
                            patternText = pattern;
                        }
                    } else {
                        weight = 1;
                        patternText = pattern;
                    }
                }
                for (var j = 0; j < weight; j++) {
                    weightedPatterns.push(patternText);
                }
            }
        } else if (collection.constructor === Array) {
            for (var i = 0, len = collection.length; i < len; i++) {
                var pattern = collection[i];
                var weight = 1, isString = pattern.constructor === String;
                var patternText = "";
                if (!isString) {
                    if (pattern.weight) {
                        weight = pattern.weight;
                    }
                    if (pattern.format) {
                        patternText = pattern.format
                    }
                } else { // look for shorthand of weight:template in string
                    var indexOfColon = pattern.indexOf(":");
                    if (indexOfColon > 0) {
                        var preColonText = pattern.substring(0, indexOfColon);
                        var number = +(preColonText);
                        if (number >= 0) {
                            weight = number;
                            patternText = pattern.substring(indexOfColon + 1);
                        } else {
                            patternText = pattern;
                        }
                    } else {
                        weight = 1;
                        patternText = pattern;
                    }
                }
                for (var j = 0; j < weight; j++) {
                    weightedPatterns.push(patternText);
                }
            }
        }
        return weightedPatterns;
    }
}
globalThis.NameGenerator = NameGenerator;

File: core/name_generator.py
"""
Procedural Name Generator for Traveller cultures and starships.
Loads template dictionaries from data/names.jsonc and evaluates token templates.
"""
from pathlib import Path
import re
from typing import Any, Callable, Dict, List, Optional
import json5
from core.dice import DiceRoller, get_roller

DATA_DIR = Path(__file__).resolve().parent.parent / "data"

def add_caps(text: str) -> str:
    """Capitalize words separated by spaces or hyphens."""
    result = []
    last_char = " "
    for char in text:
        if last_char in (" ", "-"):
            result.append(char.upper())
        else:
            result.append(char)
        last_char = char
    return "".join(result)

class NameGenerator:
    def __init__(self, data_path: Optional[Path] = None, seed: Optional[str] = None):
        if data_path is None:
            data_path = DATA_DIR / "names.jsonc"
        with open(data_path, "r", encoding="utf-8") as f:
            raw_data = json5.load(f)
        
        self.roller = get_roller(seed)
        self.templates: Dict[str, List[str]] = {}
        self._build_templates(raw_data)
        self.forbidden_words = ["bitch", "anus", "ass"]

    def set_seed(self, seed: str):
        self.roller = get_roller(seed)

    def _build_templates(self, obj: Dict[str, Any]):
        for key, collection in obj.items():
            if key == "patterns":
                continue
            self.templates[key] = self._extract_weighted_patterns(collection)
            if isinstance(collection, dict):
                for subkey, subcollection in collection.items():
                    if subkey == "patterns":
                        continue
                    self.templates[f"{key}.{subkey}"] = self._extract_weighted_patterns(subcollection)
                    if isinstance(subcollection, dict):
                        for subsubkey, subsubcollection in subcollection.items():
                            if subsubkey == "patterns":
                                continue
                            self.templates[f"{key}.{subkey}.{subsubkey}"] = self._extract_weighted_patterns(subsubcollection)

    def _extract_weighted_patterns(self, collection: Any) -> List[str]:
        weighted: List[str] = []
        patterns = []
        if isinstance(collection, dict) and "patterns" in collection:
            patterns = collection["patterns"]
        elif isinstance(collection, list):
            patterns = collection

        for pattern in patterns:
            weight = 1
            pattern_text = ""
            if isinstance(pattern, dict):
                weight = pattern.get("weight", 1)
                pattern_text = pattern.get("format", "")
            elif isinstance(pattern, str):
                if ":" in pattern:
                    parts = pattern.split(":", 1)
                    try:
                        weight = int(parts[0])
                        pattern_text = parts[1]
                    except ValueError:
                        pattern_text = pattern
                else:
                    pattern_text = pattern
            
            for _ in range(max(1, weight)):
                weighted.append(pattern_text)
        return weighted

    def unpack_string_template(self, template: str) -> List[List[Dict[str, Any]]]:
        if not template:
            return [[{"text": ""}]]
        
        # Split top level choices on '|'
        bracket_depth = 0
        phrase_strings = []
        curr = []
        for char in template:
            if char == "{":
                bracket_depth += 1
                curr.append(char)
            elif char == "}":
                bracket_depth -= 1
                curr.append(char)
            elif char == "|" and bracket_depth == 0:
                phrase_strings.append("".join(curr))
                curr = []
            else:
                curr.append(char)
        if curr:
            phrase_strings.append("".join(curr))

        # For each phrase, extract segments
        phrases: List[List[Dict[str, Any]]] = []
        for phrase_str in phrase_strings:
            segments: List[Dict[str, Any]] = []
            is_text = True
            is_ref = False
            is_ref_array = False
            curr_str = []
            ref_arr = []

            for char in phrase_str:
                if char == "{":
                    if curr_str and is_text:
                        segments.append({"text": "".join(curr_str)})
                    is_text = False
                    is_ref = True
                    curr_str = []
                elif char == "}":
                    if is_ref:
                        segments.append({"reference": "".join(curr_str)})
                        is_ref = False
                        curr_str = []
                    elif is_ref_array:
                        ref_arr.append("".join(curr_str))
                        segments.append({"references": list(ref_arr)})
                        is_ref_array = False
                        ref_arr = []
                        curr_str = []
                elif char == "|":
                    if is_ref:
                        is_ref_array = True
                        is_ref = False
                        ref_arr.append("".join(curr_str))
                        curr_str = []
                    elif is_ref_array:
                        ref_arr.append("".join(curr_str))
                        curr_str = []
                else:
                    if not is_ref and not is_ref_array:
                        is_text = True
                    curr_str.append(char)
            if curr_str and is_text:
                segments.append({"text": "".join(curr_str)})
            phrases.append(segments)
        return phrases

    def get_random_name(self, key: str, banned_words: Optional[List[str]] = None, top_level: bool = True) -> str:
        if banned_words is None:
            banned_words = self.forbidden_words
        
        templates = self.templates.get(key)
        if not templates:
            return f"Invalid Key='{key}'"

        template_roll = int(len(templates) * self.roller.random())
        template = templates[template_roll]
        phrases = self.unpack_string_template(template)
        
        phrase_roll = int(len(phrases) * self.roller.random())
        phrase = phrases[phrase_roll]

        current_text = []
        references = []
        check_not = -1

        for segment in phrase:
            if "text" in segment:
                piece = segment["text"]
                if piece.endswith(">") and references:
                    matches = re.findall(r"<!\d>", piece)
                    if matches:
                        check_not = int(re.search(r"\d", matches[-1]).group(0))
                        piece = re.sub(rf"<!{check_not}>", "", piece)
                    else:
                        check_not = -1
                else:
                    check_not = -1
                current_text.append(piece)
            elif "reference" in segment:
                ref_key = segment["reference"]
                piece = self.get_random_name(ref_key, banned_words, top_level=False)
                if check_not >= 0 and references:
                    while piece == references[min(check_not, len(references) - 1)]:
                        piece = self.get_random_name(ref_key, banned_words, top_level=False)
                current_text.append(piece)
                references.append(piece)
            elif "references" in segment:
                ref_choices = segment["references"]
                ref_key = ref_choices[int(len(ref_choices) * self.roller.random())]
                piece = self.get_random_name(ref_key, banned_words, top_level=False)
                if check_not >= 0 and references:
                    while piece == references[min(check_not, len(references) - 1)]:
                        piece = self.get_random_name(ref_key, banned_words, top_level=False)
                current_text.append(piece)
                references.append(piece)

        result_str = "".join(current_text)
        if "<" in result_str and "&" in result_str and ">" in result_str:
            for i, ref in enumerate(references):
                result_str = re.sub(rf"<&{i}>", ref, result_str)

        if result_str in banned_words and top_level:
            return self.get_random_name(key, banned_words, top_level=True)

        return add_caps(re.sub(r"\s+", " ", result_str).strip())
